src/libraries/Microsoft.Extensions.Configuration/src/ReferenceCountedProvidersManager.cs

```diff
@@ -0,0 +1,86 @@
+        private bool _disposed;
+
+        // This is only used to support IConfigurationRoot.Providers because we cannot track the lifetime of that reference.
+        public IEnumerable<IConfigurationProvider> NonReferenceCountedProviders => _refCountedProviders.NonReferenceCountedProviders;
+
+        public ReferenceCountedProviders GetReference()
+        {
+            // Lock to ensure oldRefCountedProviders.Dispose() in ReplaceProviders() doesn't decrement ref count to zero
+            // before calling _refCountedProviders.AddReference().
+            lock (_replaceProvidersLock)
+            {
+                _refCountedProviders.AddReference();
+                return _refCountedProviders;
+            }
+        }
+
+        // Providers should never be concurrently modified. Reading during modification is allowed.
+        public void ReplaceProviders(List<IConfigurationProvider> providers)
+        {
+            ReferenceCountedProviders oldRefCountedProviders = _refCountedProviders;
+
+            lock (_replaceProvidersLock)
+            {
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(nameof(ConfigurationManager));
+                }
+
+                _refCountedProviders = ReferenceCountedProviders.Create(providers);
+            }
+
+            // Decrement the reference count to the old providers. If they are being concurrently read from
+            // the actual disposal of the old providers will be delayed until the final reference is released.
+            oldRefCountedProviders.Dispose();
+        }
+
+        public void AddProvider(IConfigurationProvider provider)
+        {
+            lock (_replaceProvidersLock)
+            {
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(nameof(ConfigurationManager));
+                }
+
+                // Maintain existing references, but replace list with copy containing new item.
+                _refCountedProviders.Providers = new List<IConfigurationProvider>(_refCountedProviders.Providers)
+                {
+                    provider
+                };
+            }
+        }
+
+        public void Dispose()
+        {
+            ReferenceCountedProviders oldRefCountedProviders = _refCountedProviders;
+
+            lock (_replaceProvidersLock)
+            {
+                _disposed = true;
+
+                // Create a non-reference-counting ReferenceCountedProviders instance now that the ConfigurationManager is disposed.
+                // We could preemptively throw an ODE from ReferenceCountedProviderManager.GetReference() instead, but this might
+                // break existing apps that were previously able to continue to read configuration after disposing an ConfigurationManager.
+                _refCountedProviders = ReferenceCountedProviders.CreateDisposed(oldRefCountedProviders.Providers);
```

Comment:
Is it reasonable to delay this allocation? That way in a "normal" app that doesn't access a disposed ConfigurationManager, we don't need to allocate this one extra object.

If it is too much complexity, I'd be OK with leaving it as-is. Just needed to ask the question.

Reply:
I updated it to return a new instance every time a disposed ConfigurationManager is accessed. We could cache it, but I don't think it's worth even a little extra complexity to optimize the scenario.